### ISSUE TITLE
Configurable to allow Caps to be mapped to other things

### ISSUE THREAD

**Issue by jasonm23** [oblitum/caps2esc#4]
It would be nice if we could map Caps to be other things too.

Obviously this may be beyond your intended scope as these mappings already satisfy a common Vim use case (especially when migrating from Karabiner/OS X)

Anyway I thought I would ask the question before deciding to fork or not.

Thanks for building this.

**Comment by oblitum** [comment]
As [I've commented here](https://github.com/oblitum/caps2esc/issues/2#issuecomment-257485110), I'm working on a more general tool where caps2esc will be just an special case. I'll leave this open until I finish the other tool.

For your information, it's gonna be more or less like this:

`intercept -g $DEVICE | caps2esc | other_mapping | uinput -d $DEVICE`

`intercept` will be a tool to capture input from a given $DEVICE (grabbing it or not through the `-g` flag) and output it to `stdout`. `caps2esc` and `other_mapping` will be responsible solely for reading input events from `stdin` and outputting mutated events (or not) to `stdout`. `uinput` will be a tool to capture events from `stdin` and emulating them through a virtual device which can be constructed based on existing ones (`-d` flag) or through configuration files.

`interception` will be another tool that will accept a configuration file and will just monitor devices and execute any command when a given attached device matches the configuration. And _any command_ can be one like the one above.

**Comment by oblitum** [comment]
Just for your information, it's gonna be more or less like this:

`intercept -g $DEVICE | caps2esc | other_mapping | uinput -d $DEVICE`

`intercept` will be a tool to capture input from a given $DEVICE (grabbing it or not through the `-g` flag) and output it to `.stdout`. `caps2esc` and `other_mapping` will be responsible solely for reading input events from `stdin` and outputting mutated events or not to `stdout`. `uinput` will be a tool to capture events from `stdin` and emulating them through a virtual device that can be constructed based on existing ones or through configuration files.

`interception` will be another tool that will accept a configuration file and will just monitor devices and execute any command when a given attached device matches. And any command can be one like the one above.

**Comment by Ran4** [comment]
I personally really like that caps2esc does one thing, and one thing only. Any other tools would be cool, of course, but caps2esc staying the same would be nice.

**Comment by oblitum** [comment]
@Ran4 I agree. When I go about publishing the general tool I will provide an alternative caps2esc while this one will still live on its own.

**Comment by dragon788** [comment]
I look forward to your new tool, I'm finding that it grabbing all the xinput items without a filter is conflicting with the xinput-toggle utility that I need for another purpose.

**Comment by oblitum** [comment]
@dragon788 `caps2esc` is now superseded by this more general project https://gitlab.com/interception/linux/tools, you can use it already to filter specific devices by name or other characteristics. A more generalized `caps2esc`plugin regarding key configurations will come later.

**Comment by silenc3r** [comment]
I like caps2esc being simple but I would prefer to have Esc work as Esc and have CapsLock toggle on Ctrl. It looks like simple tweak but sadly I don't know C.

Edit: Nevermind, changing line 88 to

    if (output[0].code == KEY_LEFTCTRL)

did the trick. C ain't that hard ;).
